mops/src/hpe/cpu.tpp

```diff
@@ -2,105 +2,282 @@
 #include <algorithm>
 #include <stdexcept>
 #include <string>
-#include <iostream>
+#include <vector>
+#include <array>
 
 #include "mops/hpe.hpp"
+#include "mops/checks.hpp"
+#include "mops/utils.hpp"
 
-template<typename scalar_t>
-void mops::homogeneous_polynomial_evaluation(
-    Tensor<scalar_t, 1> output,
-    Tensor<scalar_t, 2> A,
-    Tensor<scalar_t, 1> C,
-    Tensor<int32_t, 2> indices_A
-) {
-
-    if (A.shape[0] != output.shape[0]) {
-        throw std::runtime_error(
-            "A and O tensors must have the same number of elements along the "
-            "first dimension, got " + std::to_string(A.shape[0]) + " and " +
-            std::to_string(output.shape[0])
-        );
-    }
 
-    if (C.shape[0] != indices_A.shape[0]) {
-        throw std::runtime_error(
-            "C and P tensors must have the same number of elements along the "
-            "first dimension, got " + std::to_string(C.shape[0]) + " and " +
-            std::to_string(indices_A.shape[0])
-        );
-    }
+template<typename scalar_t, uint8_t polynomial_order>
+void _homogeneous_polynomial_evaluation_templated_polynomial_order(
+    mops::Tensor<scalar_t, 1> output,
+    mops::Tensor<scalar_t, 2> A,
+    mops::Tensor<scalar_t, 1> C,
+    mops::Tensor<int32_t, 2> indices_A
+) {
 
     scalar_t* o_ptr = output.data;
-    scalar_t* a_ptr = A.data;
     scalar_t* c_ptr = C.data;
-    int32_t* indices_A_ptr = indices_A.data;
-
-    long size_batch_dimension = A.shape[0];
-    long n_monomials = indices_A.shape[0];
-    long polynomial_order = indices_A.shape[1];
-    long n_possible_factors = A.shape[1];
-
-    for (long i = 0; i < size_batch_dimension; i++) {
-        scalar_t result = 0.0;
-        scalar_t* shifted_a_ptr = a_ptr + i*n_possible_factors;
-        int32_t* indices_A_ptr_row = indices_A_ptr;
-        for (long j = 0; j < n_monomials; j++) {
-            scalar_t temp = c_ptr[j];
-            for (uint8_t k = 0; k < polynomial_order; k++) {
-                temp *= shifted_a_ptr[indices_A_ptr_row[k]];
+    int32_t* indices_a_ptr = indices_A.data;
+
+    size_t size_first_dimension = A.shape[0];
+    size_t n_monomials = indices_A.shape[0];
+    size_t n_possible_factors = A.shape[1];
+
+    constexpr size_t simd_element_count = get_simd_element_count<scalar_t>();
+
+    size_t size_first_dimension_interleft = size_first_dimension / simd_element_count;
+    size_t size_remainder = size_first_dimension % simd_element_count;
+
+    scalar_t* interleft_a_ptr = new scalar_t[size_first_dimension_interleft*n_possible_factors*simd_element_count];
+    scalar_t* remainder_a_ptr = new scalar_t[size_remainder*n_possible_factors];
```

Comment:
Instead of using `new[]`/`delete[]`; this should use `std::vector` and then `interleft_a_ptr = interleft_a.data()` if needed